File: internal/games/domain/game.go
package domain

import (
	"context"
	"time"
)

type Game interface {
	Metadata() GameMetadata
	CreateGame(ctx context.Context, input CreateGameInput) (*GameSession, error)
	ValidateGame(ctx context.Context, session *GameSession) error
	CheckResult(ctx context.Context, session *GameSession, input CheckResultInput) (*GameResult, error)
	Hint(ctx context.Context, session *GameSession, input HintInput) (*HintResult, error)
}

type GameMetadata struct {
	ID          string       `json:"id"`
	Name        string       `json:"name"`
	Description string       `json:"description"`
	Tags        []string     `json:"tags"`
	Difficulty  string       `json:"difficulty"`
	ViewKind    string       `json:"viewKind"`
	InputSchema []InputField `json:"inputSchema"`
	Defaults    CreateConfig `json:"defaults"`
}

type InputField struct {
	Name        string   `json:"name"`
	Label       string   `json:"label"`
	Type        string   `json:"type"`
	Required    bool     `json:"required"`
	Options     []string `json:"options,omitempty"`
	Placeholder string   `json:"placeholder,omitempty"`
}

type CreateConfig map[string]any

type CreateGameInput struct {
	GameID string       `json:"gameId"`
	Config CreateConfig `json:"config,omitempty"`
}

type GameSession struct {
	ID           string         `json:"id"`
	GameID       string         `json:"gameId"`
	Title        string         `json:"title"`
	Prompt       string         `json:"prompt"`
	State        map[string]any `json:"state"`
	PrivateState map[string]any `json:"-"`
	CreatedAt    time.Time      `json:"createdAt"`
}

type CheckResultInput struct {
	GameID    string         `json:"gameId"`
	SessionID string         `json:"sessionId"`
	Answer    map[string]any `json:"answer"`
}

type GameResult struct {
	Correct  bool           `json:"correct"`
	Message  string         `json:"message"`
	Score    int            `json:"score"`
	Expected map[string]any `json:"expected,omitempty"`
	Session  *GameSession   `json:"session,omitempty"`
}

type HintInput struct {
	Level int `json:"level"`
}

type HintResult struct {
	Message string `json:"message"`
	Cost    int    `json:"cost"`
}

File: swift-app/Sources/GamesApp/Domain/Game.swift
import Foundation

enum GameError: Error, Equatable {
    case gameNotFound
    case invalidSession
    case invalidAnswer
}

enum InputFieldType: String, Equatable {
    case select
    case number
    case text
}

struct InputField: Identifiable, Equatable {
    let id: String
    let label: String
    let type: InputFieldType
    let required: Bool
    let options: [String]
    let placeholder: String
}

struct GameMetadata: Identifiable, Equatable {
    let id: String
    let name: String
    let description: String
    let tags: [String]
    let difficulty: String
    let viewKind: String
    let inputSchema: [InputField]
    let defaults: [String: String]
}

struct GameSession: Identifiable {
    let id: String
    let gameID: String
    let title: String
    let prompt: String
    let state: GameState
    let privateState: PrivateGameState
    let createdAt: Date
}

enum GameState {
    case battleships(BattleshipsPuzzle)
    case wordle(WordlePuzzle)
    case memory(MemoryPuzzle)
}

enum PrivateGameState {
    case battleships(BattleshipsSolution)
    case wordle(WordleSolution)
    case memory(MemorySolution)
}

enum GameAnswer {
    case battleships(shipCoordinates: [String])
    case wordle(guess: String)
    case memory(firstIndex: Int, secondIndex: Int)
}

struct GameResult: Equatable {
    let id = UUID()
    let correct: Bool
    let message: String
    let score: Int
    let expected: [String]?

    static func == (lhs: GameResult, rhs: GameResult) -> Bool {
        lhs.id == rhs.id
    }
}

struct HintResult {
    let message: String
    let cost: Int
}

protocol Game {
    var metadata: GameMetadata { get }

    func createGame(config: [String: String]) throws -> GameSession
    func validate(session: GameSession) throws
    func checkResult(session: GameSession, answer: GameAnswer) throws -> GameResult
    func hint(session: GameSession, level: Int) throws -> HintResult
}
